platforms/core-execution/execution/README.md

```diff
@@ -0,0 +1,155 @@
+# Execution Engine
+
+The execution engine is the main component of the execution platform.
+Its purpose is to provide a simple, unified way of declaring _units of work_ that the engine can execute, producing their outputs safely and efficiently in a concurrent environment.[^concurrent-environment]
+To achieve this goal, the engine utilizes a host of safeguards and optimizations.
+Notably, the execution engine is not responsible for deciding which units of work require output production, nor is it concerned with scheduling executions.
+
+[^concurrent-environment]: This means the engine is fully thread-safe and ensures that parallel execution of work does not cause problems, even when multiple processes are executing units of work simultaneously.
+
+Any action with well-defined inputs and outputs, where safe execution or output reuse is required, can be implemented using the execution engine.
+Indeed, it is our aspiration in Gradle for all such work to be executed via the execution engine.
+
+It is also our aspiration for the execution engine to be independent of Gradle concepts.
+In part this is a good way to ensure wider usability in the future.
+But the primary reason for keeping the execution engine self-contained is to keep the architecture clean and the API contracts simple and understandable.
+
+## Unit of Work
+
+A **unit of work** in the context of the execution engine is defined by an **identifiable** **action** with a set of known **inputs** and expected **outputs**.
+The _action_ is defined by code that is a pure function with regard to its _inputs_.
+It produces equivalent _outputs_ for equivalent _inputs_.
+_Inputs_ and _outputs_ can be scalar values or files and directories.[^output-types]
+_Inputs_ are expected to be immutable after the execution of the unit starts.
+_Outputs_ should only be changed by the executing _action_.
+When work produces files, it is executed in a **workspace** directory only accessible to the work action as a means of **sandboxing**.[^task-workspace]
+
+[^output-types]: Currently, only file-like outputs are supported, but this is a historical limitation that can be lifted if needed.
+
+[^task-workspace]: Gradle tasks are currently not executed in an isolated workspace, but that is a legacy issue and should be addressed.
```

Comment:
:x: Please provide the link to the relevant issue so it's easier to find and to check if this section is still up to date